### ISSUE TITLE
Improve Handling of SearchPhaseExecutionException

### ISSUE THREAD

**Issue by ryanb** [karmi/retire#164]
It is possible for the user to trigger an exception by passing in an invalid search query. Take this code for example (provided in the sample Rails app template).

``` ruby
Article.search params[:q]
```

Pass in an empty string or a "!" and it will raise a `Tire::Search::SearchRequestFailed` exception because the backend ElasticSearch raises a `SearchPhaseExecutionException`. There may be other queries that trigger this.

I don't think the user should be able to trigger an exception. Perhaps if there is an error, return empty result, and add "valid?" and "error" methods to display the error. What do you think?

I'm using ElasticSearch v0.18.1, and Tire v0.3.12. Keep up the good work on the gem.

**Comment by karmi** [comment]
Hello Ryan,

I've been thinking about this a lot during the course of _Tire_ development, and think this behaviour — raising exceptions — is best for the library users to handle. Let me explain why I think so:

_Tire_ will [raise](https://github.com/karmi/tire/blob/master/lib/tire/search.rb#L74-76) the `SearchRequestFailed` exception whenever any kind of error with the search occurs: the query syntax is invalid (your example), index is missing, the node is down, etc etc. This way, the library user can decide, how she wants to handle the situation. Usually, in Ruby code, we do something like this:

``` ruby
    begin
      search = Tire.search('articles') { query { string 'love & hate' } }
      # ... do something with results ...
    rescue Tire::Search::SearchRequestFailed => e
      puts "Sorry! There has been and error with the search... The details are:",
           e.message
    end

    # But the script continues here just fine...
```

Now, in Rails, it's usually something like this:

``` ruby
    rescue_from Tire::Search::SearchRequestFailed do |error|
      # Indicate incorrect query to the user
      if error.message =~ /SearchParseException/ && params[:q]
        flash[:error] = "Sorry, your query syntax is invalid..."
      # ... handle other possible situations ...
      else
        redirect_to :back
      end
    end
```

I think this approach has many benefits:
- When test-driving _Tire_, everything fails fast and hard for the end user, and he begins to get a feel how to work with _elasticsearch_, which kind of queries trigger errors, etc. Everything is there, in plain sight.
- As a library user, you don't have to check for `valid?`, and don't have to inspect the `errors` Array – you just handle the exception the way you like, and display a meaningful message to the application user the way you like.

But, maybe I'm wrong — this is, in any case, the way I think about it.

As for the application user triggering an exception by entering an invalid [_Lucene_ query](http://lucene.apache.org/java/3_1_0/queryparsersyntax.html): I tend to think I'd rather handle that in one giang `case` block as suggested above in the `rescue_from` method. But there may be situations where this could become tedious, and you have to populate your `@results` anyway so your views don't fail `undefined method each for nil` etc.

To illustrate how hard is to decide stuff like this :), _elasticsearch_ itself suggests the [text query](http://www.elasticsearch.org/guide/reference/query-dsl/text-query.html) as the ideal “end-users facing query”, instead of the powerfull “query string” query, which supports all the crazy _Lucene_ meta characters...

To recap:
1. I think “failing fast and hard” is beneficial for the library users, because it's very transparent and literally hard to overlook :)
2. I'm open to any suggestions how to make the error handling more convenient, if there are some hard edges.

Best!,

Karel

**Comment by karmi** [comment]
Suggested trivial implementation in https://github.com/karmi/railscasts-episodes/commit/a2e45741d7e2ed82dffc88b6e2512655106f35a7

**Comment by karmi** [comment]
@ryanb: Ping, any news? Should we close this, or...?

**Comment by karmi** [comment]
Closed in karmi/railscasts-episodes@a2e4574.
